numpy/fft/_pocketfft_umath.c

```diff
@@ -0,0 +1,354 @@
+        if (step_out == elsize) {
+            memset(op, 0, (nout-nin)*elsize);
+        }
+        else {
+            for (npy_intp i = nin; i < nout; i++, op += step_out) {
+                memset(op, 0, elsize);
+            }
+        }
+    }
+}
+
+
+/*
+ * Loops calling the pocketfft code.
+ *
+ * Unfortunately, the gufunc machinery does not (yet?) allow forcing contiguous
+ * inner loop data, so we create a contiguous output buffer if needed
+ * (input gets copied to output before processing, so can be non-contiguous).
+ */
+static void
+fft_loop(char **args, npy_intp const *dimensions, npy_intp const *steps, void *func)
+{
+    char *ip = args[0], *fp = args[1], *op = args[2];
+    npy_intp n_outer = dimensions[0];
+    npy_intp si = steps[0], sf = steps[1], so = steps[2];
+    npy_intp nin = dimensions[1], nout = dimensions[2];
+    npy_intp isi = steps[3], iso = steps[4];
+    int (*cfft_function)(cfft_plan, double *, double) = func;
+    npy_intp npts = nout;
+    char *buffer = NULL;
+    int no_mem = 1;
+    npy_bool inplace = 0;
+
+    cfft_plan plan = make_cfft_plan(npts);
+    if (plan == NULL) {
+        goto fail;
+    }
+    if (iso != sizeof(npy_cdouble)) {
+        buffer = malloc(npts * sizeof(npy_cdouble));
+        if (buffer == NULL) {
+            goto fail;
+        }
+    }
+    else {
+        inplace = (ip == op && nin == nout && si == so && isi == iso);
+    }
+
+    for (npy_intp i = 0; i < n_outer; i++, ip += si, fp += sf, op += so) {
+        double fct = *(double *)fp;
+        char *op_buf = buffer == NULL? op : buffer;
+        if (!inplace) {  /* no copy needed if in-place already */
+            copy_data(ip, isi, nin, op_buf, sizeof(npy_cdouble), npts, sizeof(npy_cdouble));
+        }
+        if ((no_mem = cfft_function(plan, (double *)op_buf, fct)) != 0) {
+            break;
+        }
+        if (op_buf != op) {
+            copy_data(op_buf, sizeof(npy_cdouble), npts, op, iso, npts, sizeof(npy_cdouble));
+        }
+    }
+  fail:
+    if (buffer != NULL) {
+        free(buffer);
+    }
+    if (plan != NULL) {
+        destroy_cfft_plan(plan);
+    }
+    if (no_mem) {
+        /* TODO: Requires use of new ufunc API to indicate error return */
+        NPY_ALLOW_C_API_DEF
+        NPY_ALLOW_C_API;
+        PyErr_NoMemory();
+        NPY_DISABLE_C_API;
+    }
+    return;
+}
+
+
+
+static void
+rfft_impl(char **args, npy_intp const *dimensions, npy_intp const *steps, void *func,
+    npy_intp npts)
+{
+    char *ip = args[0], *fp = args[1], *op = args[2];
+    npy_intp n_outer = dimensions[0];
+    npy_intp si = steps[0], sf = steps[1], so = steps[2];
+    npy_intp nin = dimensions[1], nout = dimensions[2];
+    npy_intp isi = steps[3], iso = steps[4];
+    char *buffer = NULL;
+    int no_mem = 1;
+
+    rfft_plan plan = make_rfft_plan(npts);
+    if (plan == NULL) {
+        goto fail;
+    }
+    if (iso != sizeof(npy_cdouble)){
+        buffer = malloc(nout * sizeof(npy_cdouble));
```

Comment:
I suspect `nout` can be 0?  Did `malloc` like that (always).

Reply:
The python side ensures `nout >= 1`, but that doesn't mean one should not guard against it... So, I added a check.